### ISSUE TITLE
 Implement a more specific exception structure

### ISSUE THREAD

**Issue by ndm2** [cakephp/cakephp#9679]
This is a (multiple allowed):
* [ ] bug
* [x] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.x

---

Currently the core uses some very generic exceptions here and there, where the only difference will be the message text, which isn't something one should have to rely on for identifying the problem in a programmatic way. A good example for this is the SMTP transport, in a single `send()` call, there are 9 possible exceptions which are all of the generic type `SocketException`.

Personally I'm a big fan of having distinct exceptions for every type of error, as this makes identifying and responding to problems more easy. Sticking to the SMTP transport example, I'd think of exceptions like:

* `ConnectionFailedException`
* `TimeoutException`
* `ResponseErrorException`
* `AuthenticationFailedException`
* `PasswordRefusedException`
* `UsernameRefusedException`
* `InvalidAuthenticationMethodException`
* `UnrecognizedCommandException`
* etc...

Another prominent example would be every usage of `\Cake\Error\Exception\Exception`, while this may be the only exception being thrown in a specifc method, differentiating will become cumbersome again once that specific method is being invoked indirectly.

So long story short, what I'd like to suggest is to eliminate the usage of generic exceptions throughout the whole core by adding

* separate exceptions for every type of error
* per sub/package base exceptions

**Comment by markstory** [comment]
I think there needs to be a balance around every exceptional case needs a specific error class, and using generic errors. For example there are a few places where we raise exceptions in the Database/ORM packages to cover developer/pilot error. To me these feel like a great place to use generic errors as I don't see a use case where you would want to catch individual errors from `Cake\Database\Query::values()` as an example.

I'm not even sure if userland code would handle all the various SMTP related failures differently. I don't see how an application would react to username/password failures differently than authentication failure. But I'm likely missing something.

**Comment by dereuromark** [comment]
I think having less `\Exception` and more concrete ones is a good thing.
But we have to decide between low level developer errors (there it is probably even fine), and Runtime or catchable errors.
The latter sure should be more concrete.
Feel free to provide small PRs for certain namespaces/packages and we can discuss that then on the concrete examples.
In general it can be done for 3.next as throwing a less generic exception is BC afaik.
Let's try to improve the code here in small chunks and see where it goes.

**Comment by markstory** [comment]
>  In general it can be done for 3.next as throwing a less generic exception is BC afaik.

Not really. But throwing more specific exceptions is backwards compatible as long as the new exception is a subclass of the old exception.

**Comment by dereuromark** [comment]
Unless someone is checking for get_class() equals instead of instance_of().
But for a new minor sure something we can easily communicate.

**Comment by ndm2** [comment]
I guess there are more or less "valid" use cases for nearly everything, but to be honest, I'd probably have a hard time to come up with something useful for every possible situation :) With regards to the mentioned SMTP example, knowing what exactly caused the failure in the authentication process might for example be interesting to SMTP transport extensions that do further processing, like authentication fallback mechanisms, intrusion detection, etc.

I could imagine that for an interface that allows/requires dynamic creation of queries, like for example a database administration tool, it may even be useful to have the possibility to identify cases like `ORM::values()`. Not the best example, I know, but as I said it's hard to come up with cases that everyone would immediately be sold on.

I'd think this is also a question of how much possible extensibility improvements do matter, whether they maybe outweigh the inability to come up with use cases for every exception? Could having possibly too many specific exceptions hurt more than having possibly not enough?

**Comment by lorenzo** [comment]
I can definitely get behind having specific exceptions as long as they come with a nice error page explaining what they mean and what can be done about them.

**Comment by ndm2** [comment]
Sounds a bit as if that would be the factor (ie whether a specific error page for the exception makes sense) that you'd make the validity of having a more specific exception dependent on?

I do understand that my personal favourite of having a specific exception for at least every _type_ of error (which doesn't neccessarily have to mean every single throw) won't be met with wide support, but I think that going a little bit further than the error page factor as a limitation would be useful here and there, given that these pages are are more of a debugging aid, ie because there surely could be cases where an error page may not really be overly helpful, but the ability to catch/identify specific exceptions would be.

**Comment by markstory** [comment]
> because there surely could be cases where an error page may not really be overly helpful, but the ability to catch/identify specific exceptions would be.

I think that's good guideline to follow. If we can't come up with a good reason to try and catch specific errors or make an error page for them using a more generic exception class would work for me. What I'm most concerned about is having 100+ exception classes and each one is only used under very specific conditions.

**Comment by inoas** [comment]
If exceptions are redesigned wouldn't it make sense to set them up hierarchically (in terms of inheritance) so that one always knows which "domain" for instance an exception originated from (by looking at its inheritance)?

**Comment by markstory** [comment]
@inoas Yes I think that is what is being proposed.

**Comment by ndm2** [comment]
Correct, that's kind of what I was suggesting with per sub/package exceptions @ionas. Given that PHP has no exception interfaces, not every exception thrown in a package can make use of a base exception though, unless one would throw away the usage of specific built-in PHP exceptions and use custom ones only, which I think pretty much no one here would be a friend of :)

I'll start with some basic stuff once I have some more free time, and then we'll see what people think about it.

**Comment by ravage84** [comment]
:+1: to be less generic/more  specific with exceptions
:+1: to domain specific exceptions, does not necessariliy mean package specific, though
:-1: going overboard with it, but :+1: to go step by step

At first, I would like to see the new "big picture" aka the new Exception conecpt, an exception hierarchy proposal.
From there, we can take actions where needed.

**Comment by thinkingmedia** [comment]
I had to give a `412` response code and there was no exception.

Can we have exceptions for all the HTTP status codes? This would just be for the convenience of people using Cake.

**Comment by ravage84** [comment]
@thinkingmedia I like your idea. This came to my mind before, too. Those HTTP status code exceptions could be used in an app as server but also as client in userland code, e.g. in case of a HTTP API client.

**Comment by dereuromark** [comment]
Can we close this as no immediate action here has been taken or can be taken anymore anytime soon it seems?
